src/mig/modules/migoval/migoval.go

```diff
@@ -0,0 +1,227 @@
+	"fmt"
+	ovallib "github.com/ameihm0912/mozoval/go/src/oval"
+	"io"
+	"io/ioutil"
+	"mig/modules"
+	"time"
+)
+
+var stats Statistics
+
+// Various counters used to populate module statistics at the end of the
+// run.
+var counters struct {
+	startTime time.Time
+}
+
+func startCounters() {
+	counters.startTime = time.Now()
+}
+
+func endCounters() {
+	stats.OvalRuntime = time.Now().Sub(counters.startTime)
+}
+
+func init() {
+	modules.Register("oval", func() interface{} {
+		return new(Runner)
+	})
+}
+
+type Runner struct {
+	Parameters Parameters
+	Results    modules.Result
+}
+
+func (r Runner) Run(in io.Reader) (resStr string) {
+	defer func() {
+		if e := recover(); e != nil {
+			// return error in json
+			r.Results.Errors = append(r.Results.Errors, fmt.Sprintf("%v", e))
+			r.Results.Success = false
+			endCounters()
+			r.Results.Statistics = stats
+			err, _ := json.Marshal(r.Results)
+			resStr = string(err)
+			return
+		}
+	}()
+
+	startCounters()
+
+	// Read module parameters from stdin
+	err := modules.ReadInputParameters(in, &r.Parameters)
+	if err != nil {
+		panic(err)
+	}
+
+	err = r.ValidateParameters()
+	if err != nil {
+		panic(err)
+	}
+
+	ovallib.Init()
+	ovallib.SetMaxChecks(r.Parameters.MaxConcurrentEval)
+
+	e := &elements{}
+
+	if len(r.Parameters.PkgMatch.Matches) > 0 {
+		oresp := ovallib.PackageQuery(r.Parameters.PkgMatch.Matches)
+		for _, x := range oresp {
+			npi := &PkgInfo{PkgName: x.Name, PkgVersion: x.Version, PkgType: x.PkgType}
+			e.Matches = append(e.Matches, *npi)
+		}
+
+		r.Results.Success = true
+		if len(e.Matches) > 0 {
+			r.Results.FoundAnything = true
+		}
+		r.Results.Elements = e
+		endCounters()
+		r.Results.Statistics = stats
+		buf, err := json.Marshal(r.Results)
+		if err != nil {
+			panic(err)
+		}
+		resStr = string(buf)
+		return
+	} else if r.Parameters.OvalDef != "" {
+		b := bytes.NewBufferString(r.Parameters.OvalDef)
+		decoder := base64.NewDecoder(base64.StdEncoding, b)
+		gz, err := gzip.NewReader(decoder)
+		if err != nil {
+			panic(err)
+		}
+		ovalbuf, err := ioutil.ReadAll(gz)
+		if err != nil {
+			panic(err)
+		}
+
+		od, err := ovallib.ParseBuffer(string(ovalbuf))
+		if err != nil {
+			panic(err)
+		}
+		ovalresults, err := ovallib.Execute(od)
+		if err != nil {
+			panic(err)
+		}
+		for _, x := range ovalresults {
+			if !r.Parameters.IncludeFalse {
+				if x.Status == ovallib.RESULT_FALSE {
+					continue
+				}
+			}
+			nmor := &MOResult{}
+			nmor.Title = x.Title
+			nmor.Status = x.StatusString()
+			nmor.ID = x.ID
+			e.OvalResults = append(e.OvalResults, *nmor)
+		}
+
+		r.Results.Success = true
+		if len(e.OvalResults) > 0 {
+			r.Results.FoundAnything = true
+		}
+		r.Results.Elements = e
+		endCounters()
+		r.Results.Statistics = stats
+		buf, err := json.Marshal(r.Results)
+		if err != nil {
+			panic(err)
+		}
+		resStr = string(buf)
+		return
+	}
+
+	panic("no function specified")
+	return
+}
+
+func (r Runner) ValidateParameters() (err error) {
```

Comment:
Would it be possible to validate the oval definition file here? Since `ValidateParameters` is called by the clients, it would be useful to catch bogus files before they get sent to the agents.